src/clp_ffi_py/ir/native/PyDeserializer.cpp

```diff
@@ -0,0 +1,281 @@
+        )))
+    {
+        return -1;
+    }
+
+    if (false
+        == self->init(input_stream, buffer_capacity, static_cast<bool>(allow_incomplete_stream)))
+    {
+        return -1;
+    }
+
+    return 0;
+}
+
+CLP_FFI_PY_METHOD auto PyDeserializer_deserialize_log_event(PyDeserializer* self) -> PyObject* {
+    return self->deserialize_log_event();
+}
+
+CLP_FFI_PY_METHOD auto PyDeserializer_dealloc(PyDeserializer* self) -> void {
+    self->clean();
+    Py_TYPE(self)->tp_free(py_reinterpret_cast<PyObject>(self));
+}
+}  // namespace
+
+auto PyDeserializer::init(
+        PyObject* input_stream,
+        Py_ssize_t buffer_capacity,
+        bool allow_incomplete_stream
+) -> bool {
+    m_allow_incomplete_stream = allow_incomplete_stream;
+    m_deserializer_buffer_reader = DeserializerBufferReader::create(input_stream, buffer_capacity);
+    if (nullptr == m_deserializer_buffer_reader) {
+        return false;
+    }
+
+    try {
+        PyDeserializer::IrUnitHandler::LogEventHandle log_event_handle
+                = [this](clp::ffi::KeyValuePairLogEvent&& kv_log_event) -> IRErrorCode {
+            return this->handle_log_event(std::move(kv_log_event));
+        };
+
+        PyDeserializer::IrUnitHandler::UtcOffsetChangeHandle trivial_utc_offset_handle
+                = []([[maybe_unused]] clp::UtcOffset, [[maybe_unused]] clp::UtcOffset
+                  ) -> IRErrorCode { return IRErrorCode::IRErrorCode_Success; };
+
+        PyDeserializer::IrUnitHandler::SchemaTreeNodeInsertionHandle
+                trivial_schema_tree_node_insertion_handle
+                = []([[maybe_unused]] clp::ffi::SchemaTree::NodeLocator) -> IRErrorCode {
+            return IRErrorCode::IRErrorCode_Success;
+        };
+
+        PyDeserializer::IrUnitHandler::EndOfStreamHandle end_of_stream_handle
+                = [this]() -> IRErrorCode { return this->handle_end_of_stream(); };
+
+        auto deserializer_result{Deserializer::create(
+                *m_deserializer_buffer_reader,
+                {std::move(log_event_handle),
+                 std::move(trivial_utc_offset_handle),
+                 std::move(trivial_schema_tree_node_insertion_handle),
+                 std::move(end_of_stream_handle)}
+        )};
+        if (deserializer_result.has_error()) {
+            PyErr_Format(
+                    PyExc_RuntimeError,
+                    cDeserializerCreateErrorFormatStr.data(),
+                    deserializer_result.error().message().c_str()
+            );
+            return false;
+        }
+        m_deserializer = new clp::ffi::ir_stream::Deserializer<PyDeserializer::IrUnitHandler>{
+                std::move(deserializer_result.value())
+        };
+    } catch (clp::TraceableException& exception) {
+        handle_traceable_exception(exception);
+        return false;
+    }
+
+    return true;
+}
+
+auto PyDeserializer::deserialize_log_event() -> PyObject* {
+    try {
+        while (false == is_stream_complete()) {
```

Comment:
Nit:
I feel using `is_stream_complete()` is more confusing then directly using m_end_of_stream_reached.

Reading the code, what i feel is that you want to keep going if the stream is "incomplete", but the intension is to keep going as long as the stream hasn't ended.

If you agree with me, maybe we don't need `is_stream_complete` anymore.

Edit: or maybe you should name it as `is_stream_completed` with the extra "d"

Reply:
Sure, using `is_stream_completed` makes sense.
Just realized we used `is_stream_completed` in core as well: https://github.com/y-scope/clp/blob/main/components/core/src/clp/ffi/ir_stream/Deserializer.hpp#L105